ImageLoadingProject/ImageLoadingProject/LargeImageController.swift

```diff
@@ -0,0 +1,93 @@
+//
+//  LargeImageController.swift
+//  ImageLoadingProject
+//
+//  Created by Руслан Ахмадеев on 29.11.2020.
+//
+
+import UIKit
+
+class LargeImageController: UIViewController {
+    
+    // MARK: - UI
+    
+    @IBOutlet weak var progressView: UIProgressView!
+    @IBOutlet weak var scrollView: UIScrollView!
+    
+    private lazy var largeImage: UIImageView = {
+        let image = UIImageView()
+        image.translatesAutoresizingMaskIntoConstraints = false
+        return image
+    }()
+    
+    
+    var imageUrl: URL? {
+        didSet {
+            loadImage()
+        }
+    }
+    
+    private func setupImageView(with data: Data) {
+        let image = UIImage(data: data)
+        largeImage.image = image
+        scrollView.addSubview(largeImage)
+        NSLayoutConstraint.activate([
+            largeImage.topAnchor.constraint(equalTo: scrollView.topAnchor),
+            largeImage.leadingAnchor.constraint(equalTo: scrollView.leadingAnchor),
+            largeImage.trailingAnchor.constraint(equalTo: scrollView.trailingAnchor),
+            largeImage.bottomAnchor.constraint(equalTo: scrollView.bottomAnchor)
+        ])
+    }
+    
+    // MARK: - Helpers
+    
+    private func handleData(from url: URL) -> Data? {
+        guard let reader = try? FileHandle(forReadingFrom: url) else { return nil }
+        let data = reader.readDataToEndOfFile()
+        return data
+    }
+    
+    private func loadImage() {
+        guard let imageUrl = imageUrl else { return }
+        
+        let configuration = URLSessionConfiguration.default
+        let queue = OperationQueue()
+        
+        let session = URLSession(
+            configuration: configuration,
+            delegate: self,
+            delegateQueue: queue)
+        
+        let task = session.downloadTask(with: imageUrl)
+        task.resume()
+    }
+}
+
+// MARK: - URLSessionDownloadDelegate
+
+extension LargeImageController: URLSessionDownloadDelegate {
+    
+    func urlSession(_ session: URLSession,
+                    downloadTask: URLSessionDownloadTask,
+                    didFinishDownloadingTo location: URL) {
+        guard let data = handleData(from: location) else { return }
+        
+        DispatchQueue.main.async { [weak self] in
+            self?.setupImageView(with: data)
+            self?.progressView.removeFromSuperview()
+        }
```

Comment:
Когда есть множественный вызов weak self внутри колбэка, лучше сделать вначале `guard let self = self else { return }`, чтобы часть функций не была не вызвана, когда часть вызвана и просто для большей читабельности.

```suggestion
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.setupImageView(with: data)
            self.progressView.removeFromSuperview()
        }
```